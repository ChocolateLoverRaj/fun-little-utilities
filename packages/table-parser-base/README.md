# Table Parser Base

Utilities for table parser packages

## Usage

### Convert an `ArrayTable` to an `ObjectTable`

```javascript
const { createObjectTable } = await import('table-parser-base')

const arrayTable = {
  headers: ['id', 'name', 'email'],
  rows: [
    [1, 'John Doe', 'john-doe@gmail.com'],
    [2, 'Peter Smith', 'petersmith22@outlook.com'],
    [3, 'Julia Jones', 'jjones778@gmail.com']
  ]
}

const objectTable = createObjectTable(arrayTable)
for await (const item of objectTable) {
  console.log(item)
}
```

**Output:**

```javascript
{ id: 1, name: 'John Doe', email: 'john-doe@gmail.com' }
{ id: 2, name: 'Peter Smith', email: 'petersmith22@outlook.com' }
{ id: 3, name: 'Julia Jones', email: 'jjones778@gmail.com' }
```

### Convert an `ObjectTable` to an `ArrayTable`

```javascript
const { createArrayTable } = await import('table-parser-base')

const objectTable = [
  { id: 1, name: 'John Doe', email: 'john-doe@gmail.com' },
  { id: 2, name: 'Peter Smith', email: 'petersmith22@outlook.com' },
  { id: 3, name: 'Julia Jones', email: 'jjones778@gmail.com' }
]

const arrayTable = await createArrayTable(objectTable)
console.log(arrayTable)
```

**Output:**

```javascript
{
  headers: [ 'id', 'name', 'email' ],
  rows: [
    [ 1, 'John Doe', 'john-doe@gmail.com' ],
    [ 2, 'Peter Smith', 'petersmith22@outlook.com' ],
    [ 3, 'Julia Jones', 'jjones778@gmail.com' ]
  ]
}
```

## License

[MIT](https://git.io/fxKXN) © [Hoàng Văn Khải](https://github.com/KSXGitHub)
