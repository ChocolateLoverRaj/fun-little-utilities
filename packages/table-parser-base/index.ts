import { zipAll } from 'iter-tools'
import AdvMapInit from 'advanced-map-initialized'

export class CellSet<Title extends string, Value> {
  constructor (
    public readonly headers: readonly Title[],
    public readonly rows: Iterable<readonly Value[]>
  ) {}
}

export type ListItem<Key extends string, Value> = {
  readonly [_ in Key | symbol]: Value | undefined
}

class UnknownColumns extends AdvMapInit<number, symbol> {
  constructor () {
    super(Map, index => Symbol('Unknown Column #' + index))
  }
}

const UNKNOWN_COLUMNS = new UnknownColumns()
export const unknownColumn = (index: number) => UNKNOWN_COLUMNS.get(index)

export class List<Title extends string, Value> implements Iterable<ListItem<Title, Value>> {
  constructor (
    private readonly cells: CellSet<Title, Value>
  ) {}

  public * [Symbol.iterator] () {
    const { headers, rows } = this.cells

    for (const row of rows) {
      const item: ListItem<Title, Value> = {} as any
      let index = 0

      for (const [key, value] of zipAll<any>(headers, row)) {
        (item as any)[key ? key : unknownColumn(index)] = value
        index += 1
      }

      yield item
    }
  }
}
