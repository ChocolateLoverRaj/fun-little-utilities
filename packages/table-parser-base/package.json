{
  "name": "table-parser-base",
  "version": "0.0.0",
  "description": "Utilities for table parser packages",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://ksxnodeapps.github.io/fun-little-utilities/",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksxnodeapps/fun-little-utilities.git"
  },
  "bugs": {
    "url": "https://github.com/ksxnodeapps/fun-little-utilities/issues"
  },
  "sideEffects": false,
  "dependencies": {
    "iter-tools": "^6.2.6",
    "advanced-map-initialized": "^0.0.4",
    "tslib": "^1.10.0",
    "@types/node": "^13.5.2"
  }
}
