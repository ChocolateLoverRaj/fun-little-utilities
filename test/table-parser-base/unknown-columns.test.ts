import { unknownColumn } from 'table-parser-base'

it('same id, same symbol', () => {
  const id = 123
  expect(unknownColumn(id)).toBe(unknownColumn(id))
})

it('different id, different symbol', () => {
  expect(unknownColumn(36)).not.toBe(unknownColumn(12))
})

it('description', () => {
  expect(unknownColumn(32).description).toBe('Unknown Column #32')
})

it('snapshot', () => {
  expect([0, 1, 2, 3, 4, 5].map(unknownColumn)).toMatchSnapshot()
})
