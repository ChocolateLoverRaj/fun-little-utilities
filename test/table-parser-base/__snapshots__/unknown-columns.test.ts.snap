// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
Array [
  Symbol(Unknown Column #0),
  Symbol(Unknown Column #1),
  Symbol(Unknown Column #2),
  Symbol(Unknown Column #3),
  Symbol(Unknown Column #4),
  Symbol(Unknown Column #5),
]
`;
